src/State.php

```diff
@@ -335,217 +330,409 @@ public function prepareInputForAdd($input)
         return $input;
     }
 
+    public function post_addItem()
+    {
+        $state_visilibity = new DropdownVisibility();
+        foreach ($this->getvisibilityFields() as $itemtype => $field) {
+            if (isset($this->input[$field])) {
+                $state_visilibity->add([
+                    'itemtype' => State::getType(),
+                    'items_id' => $this->fields['id'],
+                    'visible_itemtype'  => $itemtype,
+                    'is_visible' => $this->input[$field]
+                ]);
+            }
+        }
+
+        parent::post_addItem();
+    }
+
+    public function post_updateItem($history = true)
+    {
+        $state_visilibity = new DropdownVisibility();
+        foreach ($this->getvisibilityFields() as $itemtype => $field) {
+            if (isset($this->input[$field])) {
+                if ($state_visilibity->getFromDBByCrit(['itemtype' => self::getType(), 'items_id' => $this->input['id'], 'visible_itemtype' => $itemtype])) {
+                    $state_visilibity->update([
+                        'id' => $state_visilibity->fields['id'],
+                        'is_visible' => $this->input[$field]
+                    ]);
+                } else {
+                    $state_visilibity->add([
+                        'itemtype' => State::getType(),
+                        'items_id' => $this->fields['id'],
+                        'visible_itemtype' => $itemtype,
+                        'is_visible' => $this->input[$field]
+                    ]);
+                }
+            }
+        }
+
+        parent::post_updateItem();
+    }
 
     public function rawSearchOptions()
```

Comment:
It would be interesting to have the SO list build using a foreach loop on `$CFG_GLPI['state_types']`, but I do not have any solution yet to be able to generate stable SO IDs (they should not change over time, even when we add or remove an entry in `$CFG_GLPI['state_types']`) and that takes care about the ability for plugins to add their own types inside `$CFG_GLPI['state_types']`.

Reply:
I'm hesitant to even suggest it knowing how complex it is, but maybe it is soon time to move away from numeric search option IDs.

For static numbers for each option, you can do something similar to what I did for the placeholder dashboard data to get an integer based on a string:
(int) abs((int) hexdec(hash('xxh3', 'Computer')) % 10000) + 50000;
10,000 is the maximum number to be generated and 50,000 is the starting offset.
Conflicts could still happen, even if they would be rare though. xxh3 is very fast so I don't think it will have any noticeable effect performance.

Reply:
> I'm hesitant to even suggest it knowing how complex it is, but maybe it is soon time to move away from numeric search option IDs.

We should, but the transition may be really complex and have to be specified first.

> For static numbers for each option, you can do something similar to what I did for the placeholder dashboard data to get an integer based on a string:
(int) abs((int) hexdec(hash('xxh3', 'Computer')) % 10000) + 50000;
10,000 is the maximum number to be generated and 50,000 is the starting offset.
Conflicts could still happen, even if they would be rare though. xxh3 is very fast so I don't think it will have any noticeable effect performance.

Could you propose a PR to do something like this for the `AssetDefinition` class that already produces search options based on a loop? There is no need to handle any migration as they are too recent to be used in production. If the solution is OK, it could be applied on `State` class once current PR would be merged.

Reply:
> We should, but the transition may be really complex and have to be specified first.

My only idea now is using the implementation that is done for the new API, but it needs some time actually being used by users and have more data available through it to know how good of a solution it really is. Also needs tested/tuned for performance.

> Could you propose a PR to do something like this for the AssetDefinition class

Yes